enhancements/dns/coredns-egressfirewall-integration.md

```diff
@@ -0,0 +1,363 @@
+---
+title: coredns-egressfirewall-integration
+authors:
+  - '@arkadeepsen'
+reviewers:
+  - '@Miciah'
+  - '@danwinship'
+  - '@JoelSpeed'
+  - '@TrilokGeer'
+approvers:
+  - '@Miciah'
+  - '@danwinship'
+api-approvers:
+  - '@JoelSpeed'
+creation-date: 2023-01-31
+last-updated: 2023-01-31
+tracking-link:
+  - https://issues.redhat.com/browse/CFE-748
+see-also:
+  - N.A.
+replaces:
+  - N.A.
+superseded-by:
+  - N.A.
+---
+
+# Improve CoreDNS Integration with EgressFirewall
+
+## Summary
+
+This enhancement improves the integration of CoreDNS with EgressFirewall. With this improved
+integration, EgressFirewall will be able to better support DNS names whose IPs change dynamically
+and also will be able to provide support for wildcard DNS names.
+
+## Motivation
+
+Currently, EgressFirewall (OVN-K master) does a DNS lookup of a DNS name based on a default TTL or the
+TTL of the previous lookup. OVN-K master then updates the underlying ``AddressSet`` for the DNS name
+referenced by the corresponding ACL rule of the EgressFirewall rule containing the DNS name. However,
+if a pod, belonging to the same Namespace as that of the EgressFirewall, does a DNS lookup and is able
+to get the response before the OVN-K master then the pod will be incorrectly allowed/denied access to
+the DNS name. The integration between CoreDNS and EgressFirewall needs to be improved to avoid such a
+scenario.
+
+If an administrator wants to specifically allow/deny access to all subdomains then currently the
+administrator has to add all subdomains in the EgressFirewall rules. This becomes difficult when
+subdomains are dynamically added/removed as each one has to be added/removed individually from the
+EgressFirewall rules. Currently, wildcard DNS names are not supported in EgressFirewall. However,
+even if the support is added to EgressFirewall, the integration between CoreDNS and EgressFirewall
+needs to be improved as wildcard DNS names cannot be directly looked up through a DNS query.
```

Comment:
On the contrary, there is actually another RFE for wildcard dns here: https://issues.redhat.com/browse/RFE-1384  Even if you ignore the other RFE, there are numerous issues with wildcard DNS names in an egress firewall, and you haven't addressed them in this EP.  Because of this, I cannot approve the proposal as is.

Reply:
> there are numerous issues with wildcard DNS names in an egress firewall, and you haven't addressed them in this EP

Can you please mention the issues which are not addressed in this EP?

Regarding the separate RFE, @mcurry-rh and @Deepthidharwar would be in a better position to respond to this. cc @danwinship

Reply:
PM wanted to implement the "wildcard DNS in egress firewall" feature, and the only way to do that is to change the way that all egress firewall DNS works. Since there was already an older RFE for "change the way that all egress firewall DNS works", it seemed to make more sense to call this part of that rather than vice versa. Maybe we should have done it the other way around.

Anyway, regardless of what is linked where, the actual requested-by-PM feature here is "wildcard DNS in egress firewall", so the enhancement needs to include that functionality (in some way that doesn't cause problems, obviously). If we can't figure out any way to implement wildcard DNS in egress firewall in a good way, then the enhancement would be dropped, because there's not enough customer need for _just_ "CoreDNS/EgressFirewall integration without wildcards".

Reply:
Security issues of using DNS wildcard should be part of the evaluation of accepting the RFE, but I'll do my best:
1.  a bad actor could just keep generating new dns names when there is a wildcard option to allow any new subdomains in the rules.  In other words, for a wildcard dns you would need to list all the deny lines first, then list an allow line for the wildcard subdomain.  You can deny a.b.com and allow *.b.com, but a.b.com can just change its name or publish an alias of aaa.b.com to get around it, popping a hole in the firewall
2. the pod could go directly to an upstream nameserver for resolution, bypassing the synchronization effort of the new feature
3. the pod/application could have prior knowledge of an IP address and use that instead of a dns name

Reply:
> * a bad actor could just keep generating new dns names when there is a wildcard option to allow any new subdomains in the rules.  In other words, for a wildcard dns you would need to list all the deny lines first, then list an allow line for the wildcard subdomain.  You can deny a.b.com and allow *.b.com, but a.b.com can just change its name or publish an alias of aaa.b.com to get around it, popping a hole in the firewall

The user creating the EgressFirewall object should be aware of the matching order: https://docs.openshift.com/container-platform/4.12/networking/ovn_kubernetes_network_provider/configuring-egress-firewall-ovn.html#domain-name-server-resolution_configuring-egress-firewall-ovn

Additionally, as discussed previously from the point of view of security issues only DNS allow rules should be supported. There should first be a deny all rule then DNS allow rules: https://github.com/openshift/enhancements/pull/1335#discussion_r1101797917

> * the pod could go directly to an upstream nameserver for resolution, bypassing the synchronization effort of the new feature

DNS operator sets up the pods to use CoreDNS for resolving DNS names: https://docs.openshift.com/container-platform/4.12/networking/dns-operator.html#nw-dns-operator_dns-operator

Even if this is possible, this problem would be solved by the approach mentioned above (deny all rule followed by DNS allow rules). Thus the request to the upstream nameserver will not go through.

> * the pod/application could have prior knowledge of an IP address and use that instead of a dns name

Again, this problem can be solved by the same approach (deny all rule followed by DNS allow rules). The traffic to the IP will not go through.

Reply:
From the doc: 

> The egress firewall policy rules are evaluated in the order that they are defined, from first to last. The first rule that matches an egress connection from a pod applies. Any subsequent rules are ignored for that connection.

This means if you had a deny wildcard rule first, it would not pass to any of the subsequent rules.

Reply:
It should be a IP deny rule on `0.0.0.0/0` and not a deny wildcard DNS rule. DNS rules will not be used for deny rules as mentioned above. The example CR explains how the deny rule needs to be added: https://docs.openshift.com/container-platform/4.12/networking/ovn_kubernetes_network_provider/configuring-egress-firewall-ovn.html#egressnetworkpolicy-example_configuring-egress-firewall-ovn

It is also explained in the `Important` block in this section: https://docs.openshift.com/container-platform/4.12/networking/ovn_kubernetes_network_provider/configuring-egress-firewall-ovn.html#nw-egressnetworkpolicy-about_configuring-egress-firewall-ovn